accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1179,6 +1179,119 @@ all members of `V1, .. Vk` that are not redeclared by a declaration in _DV_
 can be invoked on a receiver of the type introduced by _DV_.*
 
 
+### Changes to type functions and types
+
+The previous sections have introduced some subtype relationships involving
+extension types. This section adds further relationships where extension
+types occur more indirectly.
+
+*For instance, earlier sections stated that every extension type is a
+subtype of the top types (like `Object?`), and that `implements` introduces
+a subtype relationship for extension types. Moreover:*
+
+Assume that `E` denotes an extension type that takes `k` type arguments.
+Corresponding to the subtype rule about covariance for classes, a rule is
+introduced which makes `E<S1 .. Sk>` a subtype of `E<T1 .. Tk>` if
+`Sj` is a subtype of `Tj` for each `j` in 1 .. k.
+
+*For example, `E<int>` is a subtype of `E<Object>` because `int` is a
+subtype of `Object`, also in the case where `E` is an extension type.*
+
+The type function `futureValueType` is updated as follows:
+
+- `futureValueType(S?) = futureValueType(S)`, for any `S`.
+- `futureValueType(Future<S>) = S`, for any `S`.
+- `futureValueType(FutureOr<S>) = S`, for any `S`.
+- `futureValueType(void) = void`.
+- `futureValueType(dynamic) = dynamic`.
+- `futureValueType(E) = S` if `E` is an extension type
+  that implements `Future<S>`.
+- Otherwise, `futureValueType(S) = Object?`, for any `S`.
+
+The Dart 1 algorithm which is used to compute the standard upper bound of
+two distinct interface types will work in the same way as it does today,
+albeit on a different superinterface graph:
+
+We introduce an interface type `Any` which declares `operator ==`,
+`hashCode`, `runtimeType`, `noSuchMethod`, `toString` with the same
+signatures as `Object` does today, and `Object` as well as `Null` have
+`Any` as a direct superinterface.
+
+`Any` is sealed. *That is, we cannot create additional direct subtypes of
+`Any` This is required for the soundness of the rule below about `Object?`
+being a supertype of `Any`.*
+
+*This basically means that we eliminate the anomaly which is created by
+saying that `Object` declares those members, but `Null` "has them too,
+by magic". It also implies that all paths from the top type via `Object`
```

Comment:
It doesn't actually remove the anomaly entirely, it just fixes it for `Null`.
It doesn't explain why union types have the same members, that's just something we say that they do, or why `dynamic` treats those members specially. All it does is introduce a *nominative type* which declares an interface that everybody must implement. But we could just say that "this is the set of members that every type must implement", without making it any particular type's responsiblity to introduce them.

Reply:
We have always used the shared members as a reasonable hypothetical rule for the interface of a union type. This would simply use that as an actual rule. Granted, this is new, technically, but it shouldn't be an alien idea.

Reply:
We don't actually use the "members from both types are available on union" rule, and there are cases where it could apply, other than `Object?`. Fx:
```dart
void foo<Z extends Future<Z>>(Z z) {
  Future<Z> f1 = z;
  f.then; // Does not exist, but both `Z` and `Future<Z>` has the same member, both from `Future<Z>`.
}
```

Reply:
Closing, this text has been deleted.